constraint/pkg/client/client.go

```diff
@@ -1,355 +1,177 @@
+// by the caller.
 type Client struct {
-	driver  drivers.Driver
+	// driver contains the Rego runtime environments to run queries against.
+	// Does not require mutex locking as Driver is threadsafe.
+	driver drivers.Driver
+	// targets are the targets supported by this Client.
+	// Assumed to be constant after initialization.
 	targets map[string]handler.TargetHandler
 
-	// mtx guards access to both templates and constraints.
-	mtx       sync.RWMutex
-	templates map[templateKey]*templateEntry
-	// TODO: https://github.com/open-policy-agent/frameworks/issues/187
-	constraints map[schema.GroupKind]map[string]*unstructured.Unstructured
-
-	AllowedDataFields []string
-}
-
-// createDataPath compiles the data destination: data.external.<target>.<path>.
-func createDataPath(target, subpath string) string {
-	subpaths := strings.Split(subpath, "/")
-	p := []string{"external", target}
-	p = append(p, subpaths...)
+	// mtx guards reading and writing data outside of Driver.
+	mtx sync.RWMutex
 
-	return "/" + path.Join(p...)
+	// templates is a map from a Template's name to its entry.
+	templates map[string]*templateClient
 }
 
-// AddData inserts the provided data into OPA for every target that can handle the data.
-// On error, the responses return value will still be populated so that
-// partial results can be analyzed.
-func (c *Client) AddData(ctx context.Context, data interface{}) (*types.Responses, error) {
-	// TODO(#189): Make AddData atomic across all Drivers/Targets.
-
-	resp := types.NewResponses()
-	errMap := make(clienterrors.ErrorMap)
-	for target, h := range c.targets {
-		handled, relPath, processedData, err := h.ProcessData(data)
-		if err != nil {
-			errMap[target] = err
-			continue
-		}
-		if !handled {
-			continue
-		}
-
-		var cache handler.Cache
-		if cacher, ok := h.(handler.Cacher); ok {
-			cache = cacher.GetCache()
-		}
-
-		// Add to the target cache first because cache.Remove cannot fail. Thus, we
-		// can prevent the system from getting into an inconsistent state.
-		if cache != nil {
-			err = cache.Add(relPath, processedData)
-			if err != nil {
-				// Use a different key than the driver to avoid clobbering errors.
-				errMap[target] = err
-
-				continue
-			}
-		}
+// CreateCRD creates a CRD from template.
+func (c *Client) CreateCRD(templ *templates.ConstraintTemplate) (*apiextensions.CustomResourceDefinition, error) {
+	if templ == nil {
+		return nil, fmt.Errorf("%w: got nil ConstraintTemplate",
+			clienterrors.ErrInvalidConstraintTemplate)
+	}
 
-		// paths passed to driver must be specific to the target to prevent key
-		// collisions.
-		driverPath := createDataPath(target, relPath)
-		err = c.driver.PutData(ctx, driverPath, processedData)
-		if err != nil {
-			errMap[target] = err
+	err := validateTemplateMetadata(templ)
+	if err != nil {
+		return nil, err
+	}
 
-			if cache != nil {
-				cache.Remove(relPath)
-			}
-			continue
-		}
+	c.mtx.RLock()
```

Comment:
Do we need the lock here?

Is all we're doing reading the target handler cache? That should be constant post start-up.

Reply:
Good catch! Done.